src/v/cluster/topic_table.h

```diff
@@ -35,11 +37,32 @@ namespace cluster {
 class topic_table {
 public:
     using delta = topic_table_delta;
+
+    struct replicable_topic {
+        model::revision_id revision;
+
+        const model::topic& get_source_topic() const {
```

Comment:
I think this might be a good case for inheritance / polymorphic method. Then we can drop the use of `std::variant` , and it might be easier to understand.

Reply:
or maybe you can just add an `optional<model::topic>` to the `topic_metadata` class. If it contains a topic, its a `non_replicable` topics source, otherwise its a `replicable` topic.